query.go

```diff
@@ -320,8 +341,31 @@ func (q *query) run() {
 
 // spawnQuery starts one query, if an available heard peer is found
 func (q *query) spawnQuery(ctx context.Context, cause peer.ID, ch chan<- *queryUpdate) {
+	var qp peer.ID
 	peers := q.queryPeers.GetSortedHeard()
-	if len(peers) == 0 {
+
+	// The next peer we select for querying should be one that's not rejected by the filter.
+	for _, p := range peers {
+		if q.dFilter != nil {
+			allowed := q.dFilter.TryAdd(p)
+			if !allowed {
+				q.queryPeers.SetState(p, qpeerset.PeerRejected)
+				continue
+			} else {
+				// TODO This is hacky. Maybe, we need a `Filter.Allow()` API ?
+				// don't let unqueried peers take up slots if they are valid.
+				q.dFilter.Remove(p)
+			}
+		}
+
+		// TODO
+		// select a peer whose ASN does not conflict with that of an inflight request so we don't
+		// query peers that eventually end up rejected.
```

Comment:
@petar That's what we started with but @aschmahmann pointed out some important problems it would create by opening up new attack vectors:

https://github.com/libp2p/go-libp2p-kad-dht/pull/664#discussion_r435717783

Reply:
@aschmahmann Yeah, I get what you mean. I'll fix this.

Reply:
@aschmahmann I tried implementing this but it's messy to get it right because synchronising the query filter and the waiting filter isn't easy. For now, we can either go with a simple "don't query a peer" with the same prefix heuristic or discuss this in more depth.